File: BOJ/1012. 유기농 배추/solution.js
const query = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const dir = [
    [0, -1],
    [1, 0],
    [0, 1],
    [-1, 0]
];

let qIdx = 1, sIdx = 2;
for (let i = 1; i <= +query[0]; i += 1) {
    const [m, n, k] = query[qIdx].split(' ').map(v => +v);
    const sliced = query.slice(sIdx, sIdx + k).map(v => v.split(' ').map(e => +e));
    qIdx = sIdx + k;
    sIdx = qIdx + 1;

    solution(m, n, sliced);
}

/**
 * 
 * @param {number} m 
 * @param {number} n 
 * @param {number[]} coords 
 */
function solution(m, n, coords) {
    const farm = Array.from(Array(n), () => Array(m).fill(0));
    let cnt = 0;

    for (let i = 0; i < coords.length; i += 1) {
        const [x, y] = coords[i];
        farm[y][x] = 1;
    }

    for (let y = 0; y < n; y += 1) {
        for (let x = 0; x < m; x += 1) {
            if (farm[y][x] === 1) {
                cnt += 1;
                bfs(farm, m, n, x, y);
            }
        }
    }

    console.log(cnt);
}

function dfs(farm, m, n, x, y) {
    if (farm[y][x] === 2 || farm[y][x] === 0) return;
    farm[y][x] = 2;

    for (let i = 0; i < 4; i += 1) {
        const [dx, dy] = dir[i];
        const cx = x + dx, cy = y + dy;

        if (cx >= 0 && cy >= 0 && cx < m && cy < n) {
            dfs(farm, m, n, cx, cy);
        }
    }
}

function bfs(farm, m, n, x, y) {
    farm[y][x] = 2;
    const q = [[x, y]];

    while (q.length) {
        const [xx, yy] = q.shift();

        for (let i = 0; i < 4; i += 1) {
            const [dx, dy] = dir[i];
            const cx = xx + dx, cy = yy + dy;

            if (cx >= 0 && cy >= 0 && cx < m && cy < n) {
                if (farm[cy][cx] === 2 || farm[cy][cx] === 0) continue;
                q.push([cx, cy]);
                farm[cy][cx] = 2;
            }
        }
    }
}

File: BOJ/1012. 유기농 배추/solution.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int dx[4] = {0, 1, 0, -1};
int dy[4] = {-1, 0, 1, 0};

void dfs(vector<vector<int>> &farm, int m, int n, int x, int y) {
    if (farm[y][x] == 2 || farm[y][x] == 0) return;
    farm[y][x] = 2;

    for (int i = 0; i < 4; i += 1) {
        int cx = x + dx[i], cy = y + dy[i];

        if (cx >= 0 && cy >= 0 && cx < m && cy < n) {
            dfs(farm, m, n, cx, cy);
        }
    }
}

void bfs(vector<vector<int>> &farm, int m, int n, int x, int y) {
    farm[y][x] = 2;
    queue<pair<int, int>> q;
    q.push(make_pair(x, y));

    while(!q.empty()) {
        int yy = q.front().second;
        int xx = q.front().first;
        q.pop();

        for (int i = 0; i < 4; i += 1) {
            int cx = xx + dx[i], cy = yy + dy[i];

            if (cx >= 0 && cy >= 0 && cx < m && cy < n) {
                if (farm[cy][cx] == 2 || farm[cy][cx] == 0) continue;
                q.push(make_pair(cx, cy));
                farm[cy][cx] = 2;
            }
        }
    }
}

int solution(int m, int n, vector<pair<int, int>> coords) {
    int cnt = 0;
    vector<vector<int>> farm(n, vector<int>(m, 0));
    for (int i = 0; i < coords.size(); i += 1) {
        farm[coords[i].second][coords[i].first] = 1;
    }

    for (int y = 0; y < n; y += 1) {
        for (int x = 0; x < m; x += 1) {
            if (farm[y][x] == 1) {
                cnt += 1;
                bfs(farm, m, n, x, y);
            }
        }
    }

    cout << cnt << '\n';
    return 0;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int tc;
    cin >> tc;
    while (tc--) {
        int m, n, k;
        cin >> m >> n >> k;

        vector<pair<int, int>> coords;
        while (k--) {
        int x, y;
            cin >> x >> y;
            coords.push_back(make_pair(x, y));
        }
        solution(m, n, coords);
    }

    return 0;
}